Remove related-property mechanism and rename deferred properties to logical group properties

anttijk opened the pull request:
#### f91f78d77fdb532f75393779cbb34bf301fcb73b
<pre>
Remove related-property mechanism and rename deferred properties to logical group properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=277623">https://bugs.webkit.org/show_bug.cgi?id=277623</a>
<a href="https://rdar.apple.com/133203218">rdar://133203218</a>

Reviewed by Tim Nguyen.

The related-property mechanism has no clients left. It is also the only user of
&quot;deferred properties&quot; besides logical group properties. We can now remove this confusing
non-spec term from the codebase.

* Source/WebCore/css/process-css-properties.py:
(StylePropertyCodeGenProperties):
(StylePropertyCodeGenProperties.from_json):
(StylePropertyCodeGenProperties.is_logical):
(StyleProperty.perform_fixups_for_longhands):
(StyleProperty.perform_fixups):
(StyleProperties._sort_by_descending_priority_and_name):
(DescriptorCodeGenProperties.__init__):
(GenerateCSSPropertyNames):
(StylePropertyCodeGenProperties.is_deferred): Deleted.
(StyleProperty.perform_fixups_for_related_properties): Deleted.
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::collectMatchingRulesForList):
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::buildCascade):
(WebCore::Style::PropertyCascade::set):
(WebCore::Style::PropertyCascade::setLogicalGroupProperty):
(WebCore::Style::PropertyCascade::hasProperty):
(WebCore::Style::PropertyCascade::lastPropertyResolvingLogicalPropertyPair const):
(WebCore::Style::PropertyCascade::addMatch):
(WebCore::Style::PropertyCascade::sortLogicalGroupPropertyIDs):
(WebCore::Style::PropertyCascade::setDeferred): Deleted.
(WebCore::Style::PropertyCascade::lastDeferredPropertyResolvingRelated const): Deleted.
(WebCore::Style::PropertyCascade::sortDeferredPropertyIDs): Deleted.
* Source/WebCore/style/PropertyCascade.h:
(WebCore::Style::PropertyCascade::isEmpty const):
(WebCore::Style::PropertyCascade::hasNormalProperty const):
(WebCore::Style::PropertyCascade::logicalGroupPropertyIndex const):
(WebCore::Style::PropertyCascade::setLogicalGroupPropertyIndex):
(WebCore::Style::PropertyCascade::hasLogicalGroupProperty const):
(WebCore::Style::PropertyCascade::logicalGroupProperty const):
(WebCore::Style::PropertyCascade::logicalGroupPropertyIDs const):
(WebCore::Style::PropertyCascade::deferredPropertyIndex const): Deleted.
(WebCore::Style::PropertyCascade::setDeferredPropertyIndex): Deleted.
(WebCore::Style::PropertyCascade::hasDeferredProperty const): Deleted.
(WebCore::Style::PropertyCascade::deferredProperty const): Deleted.
(WebCore::Style::PropertyCascade::deferredPropertyIDs const): Deleted.
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyDeferredProperties):
(WebCore::Style::Builder::applyProperty):

Canonical link: <a href="https://commits.webkit.org/281834@main">https://commits.webkit.org/281834@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a1dea95d1b54e342b7a7513b537382cf4547dc37

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61169 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40530 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13750 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65119 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11718 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63299 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48207 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11993 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49455 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8159 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/63203 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37704 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52996 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30285 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34384 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10221 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10630 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56201 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10516 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66850 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5116 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10330 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56828 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5138 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52959 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57023 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13637 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4236 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36334 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37417 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38511 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37161 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
LGTM, tagging @Loirooriol since (I think?) he touched a lot of this code in case he has any ideas for cleanups.

anttijk committed:
Remove related-property mechanism and rename deferred properties to logical group properties
https://bugs.webkit.org/show_bug.cgi?id=277623
rdar://133203218

Reviewed by Tim Nguyen.

The related-property mechanism has no clients left. It is also the only user of
"deferred properties" besides logical group properties. We can now remove this confusing
non-spec term from the codebase.

* Source/WebCore/css/process-css-properties.py:
(StylePropertyCodeGenProperties):
(StylePropertyCodeGenProperties.from_json):
(StylePropertyCodeGenProperties.is_logical):
(StyleProperty.perform_fixups_for_longhands):
(StyleProperty.perform_fixups):
(StyleProperties._sort_by_descending_priority_and_name):
(DescriptorCodeGenProperties.__init__):
(GenerateCSSPropertyNames):
(StylePropertyCodeGenProperties.is_deferred): Deleted.
(StyleProperty.perform_fixups_for_related_properties): Deleted.
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::collectMatchingRulesForList):
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::buildCascade):
(WebCore::Style::PropertyCascade::set):
(WebCore::Style::PropertyCascade::setLogicalGroupProperty):
(WebCore::Style::PropertyCascade::hasProperty):
(WebCore::Style::PropertyCascade::lastPropertyResolvingLogicalPropertyPair const):
(WebCore::Style::PropertyCascade::addMatch):
(WebCore::Style::PropertyCascade::sortLogicalGroupPropertyIDs):
(WebCore::Style::PropertyCascade::setDeferred): Deleted.
(WebCore::Style::PropertyCascade::lastDeferredPropertyResolvingRelated const): Deleted.
(WebCore::Style::PropertyCascade::sortDeferredPropertyIDs): Deleted.
* Source/WebCore/style/PropertyCascade.h:
(WebCore::Style::PropertyCascade::isEmpty const):
(WebCore::Style::PropertyCascade::hasNormalProperty const):
(WebCore::Style::PropertyCascade::logicalGroupPropertyIndex const):
(WebCore::Style::PropertyCascade::setLogicalGroupPropertyIndex):
(WebCore::Style::PropertyCascade::hasLogicalGroupProperty const):
(WebCore::Style::PropertyCascade::logicalGroupProperty const):
(WebCore::Style::PropertyCascade::logicalGroupPropertyIDs const):
(WebCore::Style::PropertyCascade::deferredPropertyIndex const): Deleted.
(WebCore::Style::PropertyCascade::setDeferredPropertyIndex): Deleted.
(WebCore::Style::PropertyCascade::hasDeferredProperty const): Deleted.
(WebCore::Style::PropertyCascade::deferredProperty const): Deleted.
(WebCore::Style::PropertyCascade::deferredPropertyIDs const): Deleted.
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyDeferredProperties):
(WebCore::Style::Builder::applyProperty):

Canonical link: https://commits.webkit.org/281834@main

Loirooriol commented on Source/WebCore/css/process-css-properties.py:
```diff
@@ -959,12 +929,12 @@ def _sort_by_descending_priority_and_name(a, b):
-        a_is_deferred = a.codegen_properties.is_deferred
-        b_is_deferred = b.codegen_properties.is_deferred
-        if a_is_deferred and not b_is_deferred:
+        # Sort logical longhands to the back, before shorthands.
```
Maybe "physical/logical longhands" or "longhands in a logical property group" to be clearer.

Loirooriol commented on Source/WebCore/style/PropertyCascade.h:
```diff
@@ -124,29 +124,29 @@ class PropertyCascade {
     // 1. CSSPropertyInvalid and CSSPropertyCustom.
     // 2. Normal longhand properties (high priority ones followed by low priority ones).
-    // 3. Deferred longhand properties.
+    // 3. Logical longhand properties.
```
Ditto

Loirooriol commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -132,8 +132,8 @@ void Builder::applyNonHighPriorityProperties()
 
 void Builder::applyDeferredProperties()
```
I guess you should rename this too.
